keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,896 @@
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Pod termination by kubelet if application containers have successfully run to completion and exited.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Handling Container startup/shutdown ordering for application and sidecar containers.
+- Pod termination by kubelet if sidecar containers have exited (crash).  There will be no change in pod lifecycle if sidecar container exited.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+A container API field (I'm a keystone) can be specified by users to identify each core application containers in the pod.
+
+This identification of application containers is used by kubelet to make a decision on pod termination based on the status of application containers and pod restart policy.
+
+- if all application containers exited successfully and the restart policy of the pod is "never" or "onFailure" then terminate the pod.
+
+- if all app containers are failed and the restart policy is "never" then terminate the pod. 
+If failure logs are concerned in this case (considering the logging container is sidecar) then the sidecar container will only be terminated when all app containers are in the failed state, so it must have captured the failure logs.
+
+The following are existing behaviors just mentioned here for clarity:
+
+- if all application containers exited successfully and the restart policy of the pod is "always" then restart the containers.
+
+- if all app containers are failed and the restart policy is "onFailure", then kubelet will keep trying to restart containers, and no point in terminating the pod
+
+We propose the following API field (spec.containers[].lifecycle.type) to identify a keystone container:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+    lifecycle:
+      type: Keystone
+  - name: sidecar
+    image: sidecar-image
+    command: ["do something to help my app"]
+```
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
```

Comment:
Another use case would be "As an application that injects sidecars into pods (service mesh, logging tool, secret provider, etc), I want users of my application to not be negatively impacted by the injected sidecar or need to change their application."

The reason this is important is I think its tangibly different than the user story below. Consider I am a logging sidecar provider.

* Users already have a ton of pods or use off-the-shelf helm charts, so they aren't going to go set `keystone` on all of their pods. Especially since the person authoring the pod specs may not even know that sidecars are going to be added at all.
* Because of ^, I would think the sidecar injection process would want to automatically set the `keystone` label.
* That would work fine if there is only one sidecar injected - just add it to the other pod. But it breaks down when multiple things are injected. For example, say my webhook is called when there is 1 container in the pod. I assume it must be the keystone and inject the label. Later, another webhook is called and adds another container. This one _is_ expected to run for the entire lifetime of the pod, but because we marked the 1st container as `keystone` that won't happen.

Reply:
Not suggesting we change the whole KEP, but focusing only on this use case it would make more sense to invert the logic (eg `type: Sidecar` or equivalent).

Reply:
are there good examples of webhook injection of a container that runs to completion? Or this is a comment on breaking the existing behavior if there is such webhook exists?